Scenario sample framework (#88)
OfficeDev/office-scripts-docs: docs/resources/scenarios/analyze-web-downloads.md

---

> title: 'Office Scripts sample scenario: Internet Traffic Analysis'

Reviewer: Suggest making the title here in the metadata consistent with the H1 wording. i.e., "Internet Traffic Analysis" -> "Analyze web downloads"

description: 'A sample that takes raw internet traffic data in an Excel workbook and determines the origin location, before organizing that information into a table.'
ms.date: 02/18/2020
localization_priority: Normal
---

# Office Scripts sample scenario: Analyze web downloads

> In this scenario, you are tasked with analyzing download reports from your company's website. The goal of this analysis is to determine if the web traffic is coming from the United States or elsewhere in the world.

Reviewer: you are -> you're

> In this scenario, you are tasked with analyzing download reports from your company's website. The goal of this analysis is to determine if the web traffic is coming from the United States or elsewhere in the world.

Reviewer: should you change "United States" -> "United States (US)" here -- so that readers are clear about the meaning of the abbreviation elsewhere in the article?

> Your colleagues upload the raw data to your workbook. Each week's set of data has its own worksheet. There is also the **Summary** worksheet with a table and chart that shows week-over-week trends.

Reviewer: Consider rewording:

Your colleagues upload the raw data to your workbook. Each week's set of data has its own worksheet.

->Your colleagues upload the raw data to your workbook, where each worksheet contains one week of data.

> The script you've developed in this scenario helps analyze one of these weekly worksheets. It parses each IP address associated with a download and determines whether or not it came from the US. The answer is inserted in the worksheet as a boolean value ("TRUE" or "FALSE") and conditional formatting is applied to those cells. The IP address location results are totaled on the worksheet and copied to the summary table.

Reviewer: ```suggestion
In this scenario, you'll develop a script that analyzes weekly downloads data in the active worksheet. It'll parse the IP address associated with each download and determine whether or not it came from the US. The answer will be inserted in the worksheet as a boolean value ("TRUE" or "FALSE") and conditional formatting will be applied to those cells. The IP address location results will be totaled on the worksheet and copied to the summary table.
```

          .getRange("H2:K2");
        summaryHeaderRow.values = summaryHeader;
        summaryContentRow.values = summaryContent;
        let formats = [[".000", ".000"]];
        summaryContentRow
          .getOffsetRange(0, 2)
          .getResizedRange(0, -2).numberFormat = formats;
      }
    }

    // Translate an IP address into an integer.
    function ipInt(ip) {
      // Split the IP address into octets.
      let octets = ip.split(".");

      // Create a number for each octet and do the math to create the integer value of the IP address.
      let fullNum =
        // Define an arbitrary number for the last octet.
        111 +
        parseInt(octets[2]) * 256 +
        parseInt(octets[1]) * 65536 +
        parseInt(octets[0]) * 16777216;
      return fullNum;
    }

    // Return the row number where the ip address is found.
    function findIP(ipLookupTable: number[][], n: number) {
      for (let i = 0; i < ipLookupTable.length; i++) {
        if (ipLookupTable[i][0] <= n && ipLookupTable[i][1] >= n) {
          return i;
        }
      }
      return -1;
    }
    ```

5. Rename the script to **Analyze Web Downloads** and save it.

## Running the script

> Run the **Analyze Web Downloads** script on one of the **Week\*\*** worksheets. This will apply the conditional formatting and location labelling on the current sheet. It also updates the **Summary** worksheet.

Reviewer: Some folks may not realize that the way to "run a script on a worksheet" is to: 1) navigate to that sheet, then 2) run the script...maybe reword for clarity? 
-> `Navigate to any of the **Week\*\*** worksheets and run the **Analyze Web Downloads** script.`

> Run the **Analyze Web Downloads** script on one of the **Week\*\*** worksheets. This will apply the conditional formatting and location labelling on the current sheet. It also updates the **Summary** worksheet.

Reviewer: This will apply... -> The script will apply...

> Run the **Analyze Web Downloads** script on one of the **Week\*\*** worksheets. This will apply the conditional formatting and location labelling on the current sheet. It also updates the **Summary** worksheet.

Reviewer: It also updates... -> It'll also update

### Before running the script

![A worksheet that shows raw web traffic data.](../../images/scenario-analyze-web-downloads-before.png)

> 

Reviewer: For parity between before and after screenshots / results -- do you want to show a screenshot of the summary sheet here in the "Before" section also?

### After running the script

> ![A worksheet that shows formatted IP location information with the previous web traffic rows.](../../images/scenario-analyze-web-downloads-after.png)

Reviewer: FWIW, the "After" data looks a bit odd to me...I'd expect all the original data to be preserved and new data to be purely additive (not overwriting any of the original data). It feel weird that:
- the original value of the column H header in row 1 has been overwritten (Referrer -> Year),
- the value of column H in the first row of data (row 2) has been overwritten by the new data (http://www.... -> 2020), while original values in column H starting with the second row of data (row 3) are preserved
